### PR TITLE
[Lang] Enhance warning system and support runtime assert message

### PR DESCRIPTION
**Pull request by archibate**
Related issue = #

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

----
I don't feel so like `warnings.warn` because it can't print the full stack traceback...

### CODE REVIEW

File: python/taichi/misc/util.py [hunk 1]
```diff
@@ -132,14 +132,48 @@ def get_line_number(asc=0):
...
+    import traceback
+    import taichi as ti
+    use_spdlog = False
+    if use_spdlog:
+        s = traceback.extract_stack()[:-stacklevel]
+        raw = ''.join(traceback.format_list(s))
+        ti.warn(f'{type.__name__}: {msg}')
+        ti.warn(f'\n{raw}')
+    else:
+        warnings.warn(msg, type, stacklevel=stacklevel + 1)
```
Comment (archibate) [inline python/taichi/misc/util.py:146]: WDYT?

File: tests/python/test_assert.py [hunk 2]
```diff
@@ -26,6 +27,17 @@ def func():
     func()
 
 
+@ti.all_archs_with(debug=True, gdb_trigger=False)
```
Comment (archibate) [inline tests/python/test_assert.py:30]: #1488

**Comment by archibate** [comment]
> The code itself LGTM. One thing to think about: there will be `ti.warn` and `ti.warning`, which could be a bit confusing?

Yes, renaming one of them make sense. How about `ti.warn(msg)` -> `ti.spdlog(ti.WARN, msg)`?

**Comment by k-ye** [comment]
> How about ti.warn(msg) -> ti.spdlog(ti.WARN, msg)?

Maybe just drop `spd`, since SPD is kind of an implementation detail. 

But while I pointed out the confusion, I think it's fine to keep this as is for now. Once we have cleaned up the module relationships (e.g. your #1510) and can have nested modules, I think it's better to group these non-user facing methods in a deeper module (e.g. `ti.misc.warning` or something like that)

**Comment by archibate** [comment]
> drop spd

In fact `ti.log` is already a math function...

> #1510

Yes, we may have `ti.misc.log` and `ti.lang.log`, please ff2 leave review when you've time :)

**Pull request merged**